include/oneapi/dpl/pstl/hetero/dpcpp/unseq_backend_sycl.h

```diff
@@ -927,47 +1141,202 @@ struct __brick_includes
+            std::true_type{}, __left_start_idx,
+            oneapi::dpl::__par_backend_hetero::__scalar_store_transform_op<oneapi::dpl::__internal::__pstl_assign>{},
+            __rng_right_vector, __rng);
+    }
+    template <typename _IsFull>
+    void
+    __scalar_path_impl(_IsFull, const std::size_t __idx, _Range __rng) const
     {
         using ::std::swap;
-        swap(__acc[__idx], __acc[__size - __idx - 1]);
+        swap(__rng[__idx], __rng[__size - __idx - 1]);
+    }
+    template <typename _IsFull>
+    void
+    operator()(_IsFull __is_full, const std::size_t __idx, _Range __rng) const
+    {
+        if constexpr (__base_t::__can_vectorize)
+            __vector_path_impl(__is_full, __idx, __rng);
+        else
+            __scalar_path_impl(__is_full, __idx, __rng);
     }
 };
 
 //------------------------------------------------------------------------
 // reverse_copy
 //------------------------------------------------------------------------
-template <typename _Size>
-struct __reverse_copy
+template <typename _Size, typename _Range1, typename _Range2>
+struct __reverse_copy : public walk_vector_or_scalar_base<_Range1, _Range2>
 {
+  private:
+    using __base_t = walk_vector_or_scalar_base<_Range1, _Range2>;
+    using _ValueType = oneapi::dpl::__internal::__value_t<_Range1>;
     _Size __size;
-    template <typename _Idx, typename _AccessorSrc, typename _AccessorDst>
+
+  public:
+    __reverse_copy(_Size __size) : __size(__size) {}
+
+    template <typename _IsFull>
+    void
+    __scalar_path_impl(_IsFull, const std::size_t __idx, const _Range1 __rng1, _Range2 __rng2) const
+    {
+        __rng2[__idx] = __rng1[__size - __idx - 1];
+    }
+    template <typename _IsFull>
+    void
+    __vector_path_impl(_IsFull __is_full, const std::size_t __idx, const _Range1 __rng1, _Range2 __rng2) const
+    {
+        std::size_t __n = __size;
+        std::size_t __remaining_elements = __n - __idx;
+        std::uint8_t __elements_to_process =
+            std::min(static_cast<std::size_t>(__base_t::__preferred_vector_size), __remaining_elements);
+        const std::size_t __output_start = __size - __idx - __elements_to_process;
+        // 1. Load vector to reverse
+        _ValueType __rng1_vector[__base_t::__preferred_vector_size];
+        oneapi::dpl::__par_backend_hetero::__vector_load<__base_t::__preferred_vector_size>{__n}(
+            __is_full, __idx, oneapi::dpl::__par_backend_hetero::__scalar_load_op{}, __rng1, __rng1_vector);
+        // 2. Reverse in registers
+        oneapi::dpl::__par_backend_hetero::__vector_reverse<__base_t::__preferred_vector_size>{}(
+            __is_full, __elements_to_process, __rng1_vector);
+        // 3. Flip the location of the vector in the output buffer
+        if constexpr (_IsFull::value)
+        {
+            oneapi::dpl::__par_backend_hetero::__vector_store<__base_t::__preferred_vector_size>{__n}(
```

Comment:
is there a case when another value (not `__base_t::__preferred_vector_size`) is passing to `__vector_store` as a compile parameter?

Reply:
No, for all vector operations the preferred vector size is used.

Reply:
In that case probably it makes sense to encapsulate `__preferred_vector_size` in `__vector_store` ?

Reply:
The issue is that `__preferred_vector_size` cannot be inferred from the ranges alone which would allow us to encapsulate it into the brick without needing to pass through a template. Otherwise, I think it would be possible.